File: C++/代码随想录/Binary_Tree/Modification and construction of Binary tree/617. Merge Two Binary Trees.cpp
// 617. Merge Two Binary Trees
// https://leetcode.com/problems/merge-two-binary-trees/
// clang++ 617.\ Merge\ Two\ Binary\ Trees.cpp -std=c++17
#include "../../printUtils.hpp"
/*
这个确实简洁，但是对内存的处理不好，或者说我还不会处理这种内存。
template <typename T>
TreeNode<T>* mergeTrees(TreeNode<T>* t1, TreeNode<T>* t2) {
    if (t1 == nullptr) return t2;
    if (t2 == nullptr) return t1;
    t1->val += t2->val;
    t1->left = mergeTrees(t1->left, t2->left);
    t1->right = mergeTrees(t1->right, t2->right);

    return t1;
}
*/
template <typename T>
TreeNode<T>* mergeTrees(TreeNode<T>* t1, TreeNode<T>* t2) {
    if (t1 == nullptr && t2 == nullptr) 
        return nullptr;

    T val1 = (t1 == nullptr) ? 0 : t1->val;
    T val2 = (t2 == nullptr) ? 0 : t2->val;
    TreeNode<T>* root = new TreeNode<T>(val1 + val2);

    root->left = mergeTrees((t1 == nullptr) ? nullptr : t1->left, (t2 == nullptr) ? nullptr : t2->left);
    root->right = mergeTrees((t1 == nullptr) ? nullptr : t1->right, (t2 == nullptr) ? nullptr : t2->right);

    return root;
}

int main() {
    std::vector<int> nums1 = {1, 3, 2, 5};
    std::vector<int> nums2 = {2, 1, 3, -1, 4, -1, 7};
    TreeNode<int>* t1 = buildTree(nums1, -1);
    TreeNode<int>* t2 = buildTree(nums2, -1);

    TreeNode<int>*mergedTree = mergeTrees(t1, t2);

    std::cout << "The t1's address: " << std::endl;
    printTreeNodeAddresses(t1);

    std::cout << "The t2's address: " << std::endl;
    printTreeNodeAddresses(t2);

    std::cout << "The merged tree's address: " << std::endl;
    printTreeNodeAddresses(mergedTree);

    std::cout << "The merged tree: ";
    printArray(levelOrder(mergedTree));

    deleteTree(t1);
    deleteTree(t2);
    deleteTree(mergedTree);

    return 0;
}

File: C++/代码随想录/Binary_Tree/Modification and construction of Binary tree/617. Merge Two Binary Trees.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def mergeTrees(t1, t2):
    if not t1 and not t2:
        return None
    val1 = t1.val if t1 else 0
    val2 = t2.val if t2 else 0
    new_root = TreeNode(val1 + val2)

    new_root.left = mergeTrees(t1.left if t1 else None, t2.left if t2 else None)
    new_root.right = mergeTrees(t1.right if t1 else None, t2.right if t2 else None)

    return new_root

# 测试用例
def inorder_traversal(root):
    return inorder_traversal(root.left) + [root.val] + inorder_traversal(root.right) if root else []

# 构建测试树
t1 = TreeNode(1)
t1.left = TreeNode(3)
t1.left.left = TreeNode(5)
t1.right = TreeNode(2)

t2 = TreeNode(2)
t2.left = TreeNode(1)
t2.left.right = TreeNode(4)
t2.right = TreeNode(3)
t2.right.right = TreeNode(7)

# 合并树并输出中序遍历结果
merged_tree = mergeTrees(t1, t2)
print(inorder_traversal(merged_tree))  # [5, 4, 4, 3, 5, 7]
